### ISSUE TITLE
add cheat support

### ISSUE THREAD

**Issue by eadmaster** [joncampbell123/dosbox-x#1068]
See if [this can be of any help](https://github.com/erfg12/dosbox-cheats).

**Comment by joncampbell123** [comment]
This is the cheat system that uses pointers into DOSBox process memory directly?

No thanks.

It also doesn't take into consideration that EXE files can load to any base address, as far as I understand it.

If I were to add a cheat system to to DOSBox-X I would base it on the idea that after loading an EXE, a "patch" file can describe what bytes to patch in the resident image before running it.

Sort of like those patch files people use for NES/SNES ROM hacks but relative to the EXE load base instead of absolute memory addresses.

**Comment by joncampbell123** [comment]
I'm not sure how I'd support the same for EXE files that decompress their code at startup and then run it.

**Comment by yksoft1** [comment]
DOS world is different. Packers/encryptors or things like self-modifying code is common and they could alter, or even randomize the memory map of a game. 
Perhaps you've heard about ASLR. For some EXE encryptors, they worked like that.
RAM-based patches or GameShark-like cheat codes could work for some situations, but not all times.

**Comment by joncampbell123** [comment]
@yksoft1 Exactly. If any cheat system could be provided, the easiest would be one that patches the EXE image after loading it into memory, and that would only work for DOS games where the code or data of interest reside at a fixed memory address relative to base. It would never work with anything that randomizes the layout or decrypts/decompresses itself.

**Comment by eadmaster** [comment]
I see, a common solution for this is using **memory patterns** instead of fixed addresses to specify the cheat codes.
Once the current memory location is found it can be used as long the memory is freed.
I am not sure how easy is this to implement in dosbox, but it looks the most compatible solution to me.

**Comment by joncampbell123** [comment]
Fixed memory addresses are fine, they just have to be relative to the base of the EXE image in memory, not absolute from page zero.

**Comment by joncampbell123** [comment]
DOSBox and DOSBox-X have internal structures to keep track of which program is "current" via the current PSP segment value. The PSP segment is the first segment holding the 256-byte structure that immediately preceeds the EXE image in memory.

**Comment by eadmaster** [comment]
What if the EXE uses dynamic memory (`malloc` calls)?

**Comment by joncampbell123** [comment]
The patch system can't handle that. _But_ the code that calls malloc() is probably at a fixed location in the EXE.
